design/src/rfcs/rfc0034_generating_reusable_code.md

```diff
@@ -0,0 +1,232 @@
+        /* Deserialize ignoring the URI entirely */
+    }
+}
+```
+
+This is because `FromRequest` is parameterized by the protocol and therefore a `FromRequest<RestJson1>` can co-exist with a `FromRequest<AwsJson1_0>`.
+
+<!-- Interesting aside: If you have a recursive structure which includes an input then the code generator maps a single shape to two distinct types within one service closure. -->
+
+Why are the generated types for `RestJsonPetService > Dog > Food` and `RestJsonPetService > Cat > Food` incompatible? How should be classify such failures?
+
+### Directed Closures
+
+A (directed) closure, `closure(S)`, of a given shape `S` is all shapes visitable by downwards directed traversal. In our example model `closure(Food)` only contains `Food`, whereas `closure(RestJsonPetService)` contains everything except for `AwsJsonPetService`.
+
+The typical customer interacts with code generator by providing a service shape, via `smithy-build-template.json`, which returns the generated code for `closure(service shape)`. Everything described in this RFC is compatible with a code generator able to accept _any_ model.
+
+Models have the following noteworthy properties:
+
+- A closure of a valid model is a valid model.
+- The intersection of two valid models is a valid model.
+
+### Shape Visitor
+
+The `ShapeVisitor` interface is used as the standard entry point of the code generator. It is driven by traversal over the Smithy models ASG.
+
+```java
+public interface ShapeVisitor {
+   Object operationShape(OperationShape var1);
+
+   Object resourceShape(ResourceShape var1);
+
+   Object serviceShape(ServiceShape var1);
+
+   Object structureShape(StructureShape var1);
+
+   Object unionShape(UnionShape var1);
+
+   Object memberShape(MemberShape var1);
+
+   // ...
+}
+```
+
+On the server side, we implement this in the form of `ServerCodegenVisitor`:
+
+```kotlin
+open class ServerCodegenVisitor(
+    context: PluginContext,
+    private val codegenDecorator: ServerCodegenDecorator,
+) : ShapeVisitor.Default<Unit>()
+```
+
+## Proposal
+
+The proposal consists of a series of three principles proceeded by reasonable approach to compliance.
+
+## Principle 1) Generate Valid Code for All Valid Models
+
+> The code generator should, in principle, generate valid code for _any_ valid model.
+
+or, in terms of the `ShapeVisitor`,
+
+> Visiting any valid model should produce valid code.
+
+Notice that without complimentary principles this is not particularly strong requirement - a code generator which produces an empty crate for all models is sufficient. All proposed three principles should be used together to make decisions.
+
+## Principle 2) Only Traverse Downwards
+
+> The `ShapeVisitor` methods, if provided with access to the model, should not traverse upwards.
+
+The motivation for this principle is that adherence is sufficient to ensure the following formal description of type reusability:
```

Comment:
What I'm trying to convey here is the following hand wavey proof:

We have that `S = (S \ (S ∩ T)) ∪ (S ∩ T)` and all `S \ (S ∩ T)` are _above_ `S ∩ T`. This means that when we generate `S` and the visitor gets to `S ∩ T`, because we can't traverse upwards and inspect `S \ (S ∩ T)` we MUST act equivalently to the case where we're generating `T` (for which the same is true). Therefore it's (2) => `S ∩ T` is compatible when generating `S` and `T`.

The point of (2) is that we're not supposed to have any state concerning the parents to operate on. Maybe I should state this more explicitly? For example, having the `ShapeVisitor` record the name of the service as it traverses down the ASG is technically not a violation of (2), but is supposed to be.